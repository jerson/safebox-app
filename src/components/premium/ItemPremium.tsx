import React, {useState, useEffect} from 'react';
import {
  StyleSheet,
  ViewStyle,
  View,
  StyleProp,
  TextStyle,
  Platform,
} from 'react-native';
import Colors from '../../modules/constants/Colors';
import SplitText from '../ui/SplitText';
import Text from '../ui/Text';
import Icon from 'react-native-vector-icons/Feather';
import Client from '../../services/Client';
import {HasProductRequest, BuyProductRequest} from '../../proto/services_pb';
import Strings from '../../modules/format/Strings';
import AlertMessage from '../ui/AlertMessage';
import Button from '../ui/Button';
import {useNavigation} from 'react-navigation-hooks';
import useFocusedScreen from '../hooks/useFocusedScreen';
import * as RNIap from 'react-native-iap';
import Log from '../../modules/log/Log';

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.grey1,
  } as ViewStyle,
  content: {
    marginTop: 10,
    flexDirection: 'row',
    alignItems: 'center',
  } as ViewStyle,
  info: {
    flex: 1,
    paddingRight: 20,
  } as ViewStyle,
  description: {
    marginBottom: 10,
    fontSize: 13,
    color: Colors.grey5,
  } as TextStyle,
  icon: {
    fontSize: 28,
    marginRight: 15,
    marginLeft: 10,
    color: Colors.grey5,
  } as TextStyle,
  splitText: {
    marginBottom: 5,
  } as ViewStyle,
});

export interface ItemPremiumProps {
  style?: StyleProp<ViewStyle>;
  productId: string;
  icon: string;
  description: string;
  name: string;
  children?: React.ReactNode | React.ReactNode[];
  onPurchase?: (isPurchased: boolean) => void;
  product?: RNIap.Product;
}

const TAG = '[ItemPremium]';
function ItemPremium({
  style,
  icon,
  productId,
  description,
  name,
  onPurchase,
  product,
  children,
}: ItemPremiumProps) {
  const [isPurchased, setIsPurchased] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  const [waitForPurchase, setWaitForPurchase] = useState(false);
  const [error, setError] = useState('');

  const navigation = useNavigation();
  const [focused] = useFocusedScreen(navigation);

  useEffect(() => {
    focused && checkPurchase();
  }, [focused]);

  useEffect(() => {
    typeof onPurchase === 'function' && onPurchase(isPurchased);
  }, [isPurchased]);

  const checkPurchase = async () => {
    try {
      const request = new HasProductRequest();
      request.setSlug(productId);
      const response = await Client.hasProduct(request);
      setIsPurchased(response.getPurchased());
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }

    setIsLoading(false);
  };

  const purchase = async () => {
    setWaitForPurchase(true);
  };

  useEffect(() => {
    if (!waitForPurchase) {
      setIsLoading(false);
      return;
    }

    const purchaseUpdateSubscription = RNIap.purchaseUpdatedListener(
      async (purchase: RNIap.ProductPurchase) => {
        if (purchase.productId !== productId) {
          Log.warn(
            TAG,
            'purchaseUpdatedListener',
            'not this product',
            purchase.productId,
          );
          return;
        }
        const payload =
          purchase.purchaseToken || purchase.transactionReceipt || '';
        if (!payload) {
          setError('Try again later');
          setWaitForPurchase(false);
          Log.warn(TAG, 'purchaseUpdatedListener', 'canceled');
          return;
        }
        Log.debug(TAG, 'purchaseUpdatedListener', purchase);
        try {
          const request = new BuyProductRequest();
          request.setSlug(productId);
          request.setPayload(payload);
          request.setType(Platform.OS);
          await Client.buyProduct(request);
          setIsPurchased(true);

          if (Platform.OS === 'ios') {
            RNIap.finishTransactionIOS(purchase.transactionId || '');
          } else if (Platform.OS === 'android') {
            RNIap.acknowledgePurchaseAndroid(purchase.purchaseToken || '');
          }
        } catch (e) {
          const message = Strings.getError(e);
          setError(message);
          Log.warn(TAG, 'purchaseUpdatedListener', e);
        }

        setWaitForPurchase(false);
      },
    );

    const purchaseErrorSubscription = RNIap.purchaseErrorListener(
      (e: RNIap.PurchaseError) => {
        setError(e.debugMessage || 'Error validating purchase');
        setWaitForPurchase(false);
      },
    );
    RNIap.requestPurchase(productId, false);

    setIsLoading(true);

    const timeout = setTimeout(() => {
      setWaitForPurchase(false);
    }, 10 * 1000);
    return () => {
      purchaseErrorSubscription.remove();
      purchaseUpdateSubscription.remove();
      timeout && clearTimeout(timeout);
    };
  }, [waitForPurchase]);

  const price = product ? `- ${product.localizedPrice}` : '';
  const buttonTitle = isPurchased ? 'Purchased' : `Purchase Now ${price}`;
  return (
    <View style={[styles.container, style]}>
      <SplitText
        style={styles.splitText}
        title={`${name} - Premium`}
        type={'Default'}
      />

      {!!error && (
        <AlertMessage
          onTimeout={() => {
            setError('');
          }}
          message={error}
        />
      )}
      <View style={styles.content}>
        <Icon name={icon} style={[styles.icon]} />
        <View style={styles.info}>
          <Text style={styles.description}>{description}</Text>
          <Button
            icon={isPurchased ? undefined : 'star'}
            typeColor={'primaryLight'}
            onPress={purchase}
            disabled={isPurchased}
            isLoading={isLoading}
            title={isLoading ? undefined : buttonTitle}
          />
        </View>
      </View>
      {children}
    </View>
  );
}

export default ItemPremium;
