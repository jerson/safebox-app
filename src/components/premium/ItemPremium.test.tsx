import React from 'react';
import * as renderer from 'react-test-renderer';
import ItemPremium, {ItemPremiumProps} from './ItemPremium';

const props: ItemPremiumProps = {
  description: 'sample',
  name: 'sample',
  icon: 'trash',
  productId: 'sample',
};
test('renders without crashing', () => {
  const snapshot = renderer.create(<ItemPremium {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<ItemPremium {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
