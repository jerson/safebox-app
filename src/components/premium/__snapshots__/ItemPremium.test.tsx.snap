// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views 1`] = `
<_class
  style={
    Array [
      Object {
        "backgroundColor": "#fafdff",
      },
      undefined,
    ]
  }
>
  <_class
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "marginBottom": 5,
        },
      ]
    }
  >
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#9066f5",
            "height": 1,
            "left": 0,
            "opacity": 0.5,
            "position": "absolute",
            "right": 0,
            "top": 12,
          },
          Object {
            "backgroundColor": "#c2d0e7",
          },
          false,
        ]
      }
    />
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#4531cd",
            "borderRadius": 10,
            "paddingBottom": 4,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 6,
          },
          Object {
            "backgroundColor": "#6781a6",
          },
          false,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#FFFFFF",
              "fontSize": 13,
              "lineHeight": 14,
              "minWidth": 20,
              "textAlign": "center",
            },
          ]
        }
      >
        sample - Premium
      </Text>
    </_class>
  </_class>
  <_class
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Array [
            Object {
              "color": "#65748b",
              "fontSize": 28,
              "marginLeft": 10,
              "marginRight": 15,
            },
          ],
          Object {
            "fontFamily": "Feather",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <_class
      style={
        Object {
          "flex": 1,
          "paddingRight": 20,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#65748b",
              "fontSize": 13,
              "marginBottom": 10,
            },
          ]
        }
      >
        sample
      </Text>
      <_class
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#9066f5",
            "borderColor": "#9066f5",
            "borderRadius": 10,
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 0,
            "opacity": 1,
            "padding": 12,
            "zIndex": 1,
          }
        }
        testID="button_touchable"
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 14,
                  "marginRight": 4,
                  "top": 3,
                },
                Object {
                  "color": "#FFFFFF",
                },
                undefined,
              ],
              Object {
                "fontFamily": "Feather",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Nunito",
                "fontSize": undefined,
                "fontStyle": "normal",
                "fontWeight": "400",
              },
              Object {},
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 14,
                  "textAlign": "center",
                },
                Object {
                  "color": "#FFFFFF",
                },
                undefined,
              ],
            ]
          }
        >
          Purchase Now 
        </Text>
      </_class>
    </_class>
  </_class>
</_class>
`;
